controller/src/lib.rs

```diff
@@ -124,73 +193,158 @@ fn schedule_response_for_existing_backend(
 }
 
 async fn process_response(
-    state: &StateHandle,
+    state: &mut StateHandle,
     sr: &ScheduleRequest,
     scheduler: &Scheduler,
     nats: &TypedNats,
 ) -> anyhow::Result<ScheduleResponse> {
     tracing::info!("checking locks");
     let cluster_name = sr.cluster.clone();
-    if let Some(lock_name) = sr.lock.clone() {
+    let schedule_response: Option<_> = if let Some(lock_name) = sr.lock.clone() {
         tracing::info!(?lock_name, "scheduling lock");
 
-        if let Ok(backend) = backend_of_lock(state, &cluster_name, &lock_name) {
-            tracing::info!(?backend, "fetch preexisting backend");
-            return schedule_response_for_existing_backend(state, cluster_name, backend);
-        }
+        if let Ok(Some(backend)) = backend_assigned_to_lock(state, &lock_name, &cluster_name) {
+            Some(schedule_response_for_existing_backend(
+                state,
+                cluster_name.clone(),
+                backend,
+            ))
+        } else {
+            let mut rng = rand::rngs::StdRng::from_entropy();
+            let my_uid: u128 = rng.gen();
 
-        tracing::info!("spawn with lock");
-    }
+            let my_announce_time = announce_lock(&lock_name, &cluster_name, nats, &my_uid).await?;
+
+            state.wait_for_seq(my_announce_time).await;
+            let lock_state = state
+                .state()
+                .cluster(&cluster_name)
+                .expect("cluster should exist")
+                .locks
+                .get(&lock_name)
+                .expect("lock should be in map after waiting for announce time")
```

Comment:
I'm a bit more... free with panicking in these functions since they're all scoped within a short lived tokio::spawn, and that top level is in this file, and all it's doing is logging. but I guess panics don't get forwarded over tracing? 

yep will propagate.